feat: add WebView support for SAML registration/login flows

splaunov opened the pull request:
<!--
Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request.

This text will be included in the changelog. If applicable, include links to documentation or pieces of code.
If your change includes breaking changes please add a codeblock documenting the breaking change:

```
BREAKING CHANGES: This patch changes the behavior of configuration item `foo` to do bar. To keep the existing
behavior please do baz.
```
-->

The registration/login flow should end with special redirect if used in WebView on mobile platform

## Related issue(s)

#2653

<!--
If this pull request

1. is a fix for a known bug, link the issue where the bug was reported in the format of `#1234`;
2. is a fix for a previously unknown bug, explain the bug and how to reproduce it in this pull request;
3. implements a new feature, link the issue containing the design document in the format of `#1234`;
4. improves the documentation, no issue reference is required.

Pull requests introducing new features, which do not have a design document linked are more likely to be rejected and take on average 2-8 weeks longer to
get merged.

You can discuss changes with maintainers either in the Github Discusssions in this repository or
join the [Ory Chat](https://www.ory.sh/chat).
-->

## Checklist

<!--
Put an `x` in the boxes that apply. You can also fill these out after creating the PR.

Please be aware that pull requests must have all boxes ticked in order to be merged.

If you're unsure about any of them, don't hesitate to ask. We're here to help!
-->

- [x ] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md).
- [x ] I have referenced an issue containing the design document if my change introduces a new feature.
- [x ] I am following the [contributing code guidelines](../blob/master/CONTRIBUTING.md#contributing-code).
- [x ] I have read the [security policy](../security/policy).
- [x ] I confirm that this pull request does not address a security vulnerability. 
      If this pull request addresses a security. vulnerability, 
      I confirm that I got green light (please contact [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push the changes.
- [x ] I have added tests that prove my fix is effective or that my feature works.
- [ ] I have added or changed [the documentation](https://github.com/ory/docs).

## Further Comments

<!--
If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution
you did and what alternatives you considered, etc...
-->

splaunov commented:
Have added WebView support and test

Stoakes commented on selfservice/flow/registration/error.go:
```diff
@@ -115,7 +117,38 @@ func (s *ErrorHandler) WriteFlowError(
+
+			query := returnTo.Query()
+
+			query.Set("code", strconv.Itoa(int(f.UI.Messages[0].ID)))
```
Are we sure `f.UI.Messages[0]` exists ?

splaunov replied:
Have added check for messages array length.

Stoakes commented on selfservice/strategy/saml/strategy/test/strategy_helper_test.go:
```diff
@@ -0,0 +1,154 @@
+package saml_test
+
+import (
+	"context"
+	"fmt"
+	"github.com/ory/dockertest/v3"
+	"github.com/ory/dockertest/v3/docker"
+	"github.com/ory/kratos/driver"
+	"github.com/ory/kratos/driver/config"
+	"github.com/ory/kratos/identity"
+	"github.com/ory/kratos/selfservice/strategy/saml"
+	"github.com/ory/x/logrusx"
+	"github.com/ory/x/resilience"
+	"github.com/phayes/freeport"
+	"github.com/pkg/errors"
+	"github.com/stretchr/testify/require"
+	"golang.org/x/net/html"
+	"net/http"
+	"net/http/cookiejar"
+	"net/http/httptest"
+	"testing"
+	"time"
```
Can you please order the imports this way in all files ? 

```
import (
   "std libs"

  " project libs"

"external dependencies"
)
```

In the case of this file, that would be 

```suggestion
	"context"
	"fmt"
	"net/http"
	"net/http/cookiejar"
	"net/http/httptest"
	"testing"
	"time"
	
	"github.com/ory/dockertest/v3"
	"github.com/ory/dockertest/v3/docker"
	"github.com/ory/kratos/driver"
	"github.com/ory/kratos/driver/config"
	"github.com/ory/kratos/identity"
	"github.com/ory/kratos/selfservice/strategy/saml"
	"github.com/ory/x/logrusx"
	"github.com/ory/x/resilience"
	
	"github.com/phayes/freeport"
	"github.com/pkg/errors"
	"github.com/stretchr/testify/require"
	"golang.org/x/net/html"
```

splaunov replied:
Have run `make format`

Stoakes commented on selfservice/strategy/saml/strategy/test/strategy_test.go:
```diff
@@ -1,23 +1,242 @@
 
+const debugRedirects = false
+
+func TestStrategy(t *testing.T) {
```
Generally speaking, there could be more comments in this PR. To detail what test cases do, what are the expected outcome ... to have the main lines of what the code actually do.

splaunov replied:
Have added comments to key test steps

sebferrer reviewed (changes requested):
In order to better follow Ory's design pattern, we have reorganized the files concerning our part of the code. The SAML handler is now in the strategy package, no longer in flow.
Could you please follow this new file architecture in your PR?
Thanks again! :)

splaunov commented:
Yes, will update my branch.

splaunov commented:
The branch has been updated. Please review and merge.

splaunov commented:
Have added validation error for unknown provider. We need this when the provider discovery is based on a user input.

sebferrer commented:
Hey @splaunov! Sorry we've been very busy lately. We had to force-push to rebase on the ory/kratos master. The changes are not major and should not generate too many conflicts. Can you please rebase to our SAML branch? Sorry for the inconvenience. We will finally have time to take care of the merge of this PR :)

aeneasr committed:
test: resolve failing test for empty tokens (#3775)

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

jonas-jonas committed:
fix: test assertions on declassifying OIDC tokens (#3773)

Co-authored-by: aeneasr <3372410+aeneasr@users.noreply.github.com>

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

aeneasr committed:
feat: add twitter SSO (#3778)

aeneasr committed:
chore: remove e2e playwright env (#3794)

ory-bot committed:
autogen(openapi): regenerate swagger spec and internal client

[skip ci]

jonas-jonas committed:
fix: make sure emails can still be sent with SMS enabled (#3795)

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

aeneasr committed:
fix: show error page on identity mismatch (#3790)

hperl committed:
fix: audit issues (#3797)

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

hperl committed:
feat: send OIDC claim keys to tracing (#3798)

ory-bot committed:
autogen(openapi): regenerate swagger spec and internal client

[skip ci]

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

alnr committed:
fix: missing indices and foreign keys (#3800)

aeneasr committed:
feat: control edge cache ttl (#3808)

ory-bot committed:
autogen(openapi): regenerate swagger spec and internal client

[skip ci]

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

dependabot[bot] committed:
chore(deps): bump github.com/lestrrat-go/jwx from 1.2.28 to 1.2.29 (#3812)

Bumps [github.com/lestrrat-go/jwx](https://github.com/lestrrat-go/jwx) from 1.2.28 to 1.2.29.
- [Release notes](https://github.com/lestrrat-go/jwx/releases)
- [Changelog](https://github.com/lestrrat-go/jwx/blob/v1.2.29/Changes)
- [Commits](https://github.com/lestrrat-go/jwx/compare/v1.2.28...v1.2.29)

---
updated-dependencies:
- dependency-name: github.com/lestrrat-go/jwx
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

hperl committed:
feat: PassKeys with Resident Keys and two-step registration (#3748)

BREAKING CHANGES: This feature enables two-step registration per default. Two-step registration is a significantly improved sign up flow and recommended when using more than one sign up methods. To disable two-step registration, set `selfservice.flows.registration.enable_legacy_flow` to `true`. This value defaults to `false`.

ory-bot committed:
autogen(openapi): regenerate swagger spec and internal client

[skip ci]

dependabot[bot] committed:
chore(deps): bump github.com/go-jose/go-jose/v3 from 3.0.1 to 3.0.3 (#3805)

Bumps [github.com/go-jose/go-jose/v3](https://github.com/go-jose/go-jose) from 3.0.1 to 3.0.3.
- [Release notes](https://github.com/go-jose/go-jose/releases)
- [Changelog](https://github.com/go-jose/go-jose/blob/v3.0.3/CHANGELOG.md)
- [Commits](https://github.com/go-jose/go-jose/compare/v3.0.1...v3.0.3)

---
updated-dependencies:
- dependency-name: github.com/go-jose/go-jose/v3
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

hperl committed:
feat: linkedin v2 provider (#3804)

* feat: add linkedin-v2 provider

* docs: document linkedin special-case

ory-bot committed:
autogen(openapi): regenerate swagger spec and internal client

[skip ci]

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

aeneasr committed:
fix(sdk): improve discriminators for node and Go (#3821)

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

aeneasr committed:
feat: use authenticate endpoint for x (#3833)

Improves the "Log in with X" experience by not asking the user to re-authenticate every time.

ory-bot committed:
autogen(openapi): regenerate swagger spec and internal client

[skip ci]

jonas-jonas committed:
feat: add verification hook to login flow (#3829)

ory-bot committed:
autogen(openapi): regenerate swagger spec and internal client

[skip ci]

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

zepatrik committed:
fix: drop trigram index on identifiers (#3827)

ory-bot committed:
autogen(openapi): regenerate swagger spec and internal client

[skip ci]

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

hperl committed:
fix: passing transient payloads (#3838)

ory-bot committed:
autogen(openapi): regenerate swagger spec and internal client

[skip ci]

ngc7293 committed:
fix: tolerate more "truthy" values when creating new flows (#3841)

Use strconv.ParseBool to accept multiple "truthy" values for the
`refresh` and `return_session_token_exchange_code` query parameters when
creating a new login flow.

For some SDKs (e.g.: Python), these stringification of booleans is not
user-controlled and these endpoints could not be used fully due to the
backend ignoring any value other than `true` (all lowercase).

Closes #3839

aeneasr committed:
fix(sdk): expand identity in session extension (#3843)

Closes #3842

aeneasr committed:
fix: improve SDK discriminators (#3844)

aeneasr committed:
chore: upgrade ory/x to v0.0.619 (#3845)

jonas-jonas committed:
fix: execute verification & verification_ui properly in login flows (#3847)

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

hperl committed:
fix: don't treat passkeys as AAL2 (#3853)

aeneasr committed:
fix: drop index if exists (#3846)

aeneasr committed:
feat: add additional messages

aeneasr committed:
fix: add missing env vars to set up guide (#3855)

Closes https://github.com/ory/kratos/issues/3828

sashatalalasha committed:
feat: add headers to web hooks (#3849)

hperl committed:
fix: webhook transient payload in OIDC login flows (#3857)

* fix: transient payload with OIDC login

ory-bot committed:
autogen(openapi): regenerate swagger spec and internal client

[skip ci]

jonas-jonas committed:
fix: don't require connection_uri in SMTP (#3861)

jonas-jonas committed:
fix: do not require method to be  passkey in settings schema (#3862)

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

aeneasr committed:
fix: use correct post-verification identity state in post-hooks (#3863)

ory-bot committed:
autogen(openapi): regenerate swagger spec and internal client

[skip ci]

hperl committed:
fix: linkedin issuer override (#3875)

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

jonas-jonas committed:
test: deflake session test (#3864)

aeneasr committed:
feat: add session to post login webhook (#3877)

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

hperl committed:
fix: always issue session last (#3876)

In post persist hooks, the session issuance hook always needs
to come last. This fixes the getHooks function to ensure this.

alnr committed:
fix: tweaks to UpsertSessions (#3878)

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

alnr committed:
fix: allow updating just the verified_at timestamp of addresses (#3880)

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

aeneasr committed:
feat: add redirect to continue_with for SPA flows

This patch adds the new `continue_with` action `redirect_browser_to`, which contains the redirect URL the app should redirect to. It is only supported for SPA (not server-side browser apps, not native apps) flows at this point in time.

hperl committed:
fix: include all creds in duplicate credential err (#3881)

ory-bot committed:
autogen(openapi): regenerate swagger spec and internal client

[skip ci]

aeneasr committed:
feat: add browser return_to continue_with action

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

jonas-jonas committed:
fix: enum type of session expandables (#3891)

ory-bot committed:
autogen(openapi): regenerate swagger spec and internal client

[skip ci]

jonas-jonas committed:
chore: render CLI doc messages into their own *.md file in docs (#3886)

aeneasr committed:
feat(sdk): add missing profile discriminator to update registration

aeneasr committed:
feat(sdk): avoid eval with javascript triggers

Using `OnLoadTrigger` and `OnClickTrigger` one can now map the trigger to the corresponding JavaScript function.

For example, trigger `{"on_click_trigger":"oryWebAuthnRegistration"}` should be translated to `window.oryWebAuthnRegistration()`:

```
if (attrs.onClickTrigger) {
  window[attrs.onClickTrigger]()
}
```

hperl committed:
fix: respect return_to in OIDC API flow error case (#3893)

* fix: respect return_to in OIDC API flow error case

This fix ensures that we redirect the user to the return_to URL
when an error occurs during the OIDC login for native flows.

Native flows are initialized through the API, and the browser
URL is retrieved from a 422 response after a POST to submit the
login flow. Successful OIDC flows already returned the `code` to
the `return_to` URL. Now, unsuccessful flows return the `flow` with
the current flow ID (which might have changed), so that the caller
can retrieve the full flow and act accordingly.

* fix: ignore trivvy CVE report

Bump in distroless is still open

ory-bot committed:
autogen(openapi): regenerate swagger spec and internal client

[skip ci]

jonas-jonas committed:
fix: enum type of session expandables (#3895)

ory-bot committed:
autogen(openapi): regenerate swagger spec and internal client

[skip ci]

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

aeneasr committed:
fix: replace submit with continue button for recovery and verification and add maxlength

aeneasr committed:
feat: set maxlength for totp input

hperl committed:
fix: db index and duplicate credentials error (#3896)

* fix: don't return password cred type if empty
* fix: better index for config.user_handle on identity_credentials

ory-bot committed:
autogen(openapi): regenerate swagger spec and internal client

[skip ci]

testwill committed:
fix: close res body (#3870)

Signed-off-by: guoguangwu <guoguangwug@gmail.com>

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

camcui committed:
chore: fix function name in comment (#3869)

Signed-off-by: camcui <cuishua@sina.cn>
Co-authored-by: Jonas Hungershausen <jonas.hungershausen@ory.sh>

jonas-jonas committed:
docs: remove delete reference from batch patch identity (#3906)

ory-bot committed:
autogen(openapi): regenerate swagger spec and internal client

[skip ci]

alnr committed:
fix: CVEs in dependencies (#3902)

aeneasr committed:
chore: make identity schema provider a proper service (#3908)

ory-bot committed:
autogen(openapi): regenerate swagger spec and internal client

[skip ci]

aeneasr committed:
feat: identifier first auth

aeneasr committed:
fix(oidc): grace period for continuity container on oidc callbacks (#3915)

ory-bot committed:
autogen(openapi): regenerate swagger spec and internal client

[skip ci]

aeneasr committed:
chore: allow smtp jim config (#3932)

jacwil committed:
fix: change return urls in quickstarts (#3928)

aeneasr committed:
chore: update repository templates to https://github.com/ory/meta/commit/e838bee8d0a29d022b61472be9efccf096099130

aeneasr committed:
fix: add missing JS triggers

aeneasr committed:
feat: add if method to sdk

aeneasr committed:
chore: regenerate SDK

aeneasr committed:
feat: identifier first login for all first factor login methods

jonas-jonas committed:
test: resolve flaky e2e tests (#3935)

* test: resolve flaky code registration tests

* chore: don't fail logout if cookie is not found

* chore: remove .only

* chore: reduce wait

* chore: u

* chore: u

* chore: u

jonas-jonas committed:
chore: upgrade nyaruka/phonenumbers to v1.3.6 (#3940)

aeneasr committed:
chore: improve courier logging (#3943)

zepatrik committed:
feat: allow admin to create API code recovery flows (#3939)

ory-bot committed:
autogen(openapi): regenerate swagger spec and internal client

[skip ci]

aeneasr committed:
autogen: pin v1.2.0-pre.0 release commit

aeneasr committed:
autogen: pin v1.2.0 release commit

ory-bot committed:
autogen(docs): generate and bump docs

[skip ci]

ory-bot committed:
autogen: add v1.2.0 to version.schema.json

[skip ci]

Demonsthere committed:
chore: add kubescape image scanner (#3947)

ory-bot committed:
autogen(openapi): regenerate swagger spec and internal client

[skip ci]

aeneasr committed:
feat: improve session extend performance (#3948)

This patch improves the performance for extending session lifespans. Lifespan extension is tricky as it is often part of the middleware of Ory Kratos consumers. As such, it is prone to transaction contention when we read and write to the same session row at the same time (and potentially multiple times).

To address this, we:

1. Introduce a locking mechanism on the row to reduce transaction contention;
2. Add a new feature flag that toggles returning 204 no content instead of 200 + session.

Be aware that all reads on the session table will have to wait for the transaction to commit before they return a value. This may cause long(er) response times on `/session/whoami` for sessions that are being extended at the same time.

BREAKING CHANGES: Going forward, the `/admin/session/.../extend` endpoint will return 204 no content for new Ory Network projects. We will deprecate returning 200 + session body in the future.

ory-bot committed:
autogen(openapi): regenerate swagger spec and internal client

[skip ci]

aeneasr committed:
chore: move b2b config to selfservice section (#3949)

ory-bot committed:
autogen(openapi): regenerate swagger spec and internal client

[skip ci]

zepatrik committed:
test: deflake session extend config side-effect (#3950)

aeneasr committed:
test: add form hydration tests for password login

aeneasr committed:
test: add form hydration tests for code login

aeneasr committed:
test: add form hydration tests for oidc login

aeneasr committed:
test: add form hydration tests for idfirst login

aeneasr committed:
feat: add social providers to credential discovery as well

aeneasr committed:
test: add form hydration tests for passkey login

aeneasr committed:
test: add form hydration tests for webauthn login

zepatrik committed:
test: deflake and parallelize persister tests (#3953)

ory-bot committed:
autogen(openapi): regenerate swagger spec and internal client

[skip ci]

aeneasr committed:
test: add tests for idfirst

aeneasr committed:
chore: updated snapshots and clean up

aeneasr committed:
test: verify redirect continue_with in hook executor for browser clients

zepatrik committed:
test: enable server-side config from context (#3954)

aeneasr committed:
test: resolve issues and update snapshots for all selfservice strategies

aeneasr committed:
feat: clarify session extend behavior (#3962)

aeneasr committed:
feat: separate 2fa refresh from 1st factor refresh (#3961)

aeneasr committed:
feat: better detection if credentials exist on identifier first login (#3963)

alnr committed:
feat: allow deletion of an individual OIDC credential (#3968)

This extends the existing `DELETE /admin/identities/{id}/credentials/{type}` API to accept an `?identifier=foobar` query parameter for `{type}==oidc` like such:

`DELETE /admin/identities/{id}/credentials/oidc?identifier=github%3A012345`

This will delete the GitHub OIDC credential with the identifier `github:012345` (`012345` is the subject as returned by GitHub).

To find out which OIDC credentials exist, call `GET /admin/identities/{id}?include_credential=oidc` beforehand.

This will allow you to delete individual OIDC credentials for users even if they have several set up.

ory-bot committed:
autogen(openapi): regenerate swagger spec and internal client

[skip ci]

David-Wobrock committed:
docs: typo in changelog

alnr committed:
fix: jsonnet timeouts (#3979)

hperl committed:
feat: password migration hook (#3978)

This adds a password migration hook to easily migrate passwords for which we do not have the hash.

For each user that needs to be migrated to Ory Network, a new identity is created with a credential of type password with a config of {"use_password_migration_hook": true} .
When a user logs in, the credential identifier and password will be sent to the password_migration web hook if all of these are true:
The user’s identity’s password credential is {"use_password_migration_hook": true}
The password_migration hook is configured
After calling the password_migration hook, the HTTP status code will be inspected:
On 200, we parse the response as JSON and look for {"status": "password_match"}. The password credential config will be replaced with the hash of the actual password.
On any other status code, we assume that the password is not valid.

---------

Co-authored-by: zepatrik <zepatrik@users.noreply.github.com>

ory-bot committed:
autogen(openapi): regenerate swagger spec and internal client

[skip ci]

jonas-jonas committed:
chore: use label in link/unlink settings nodes (#3977)

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

jonas-jonas committed:
chore: improve courier logging (#3985)

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

hperl committed:
fix: move password migration hook config (#3986)

This moves the password migration hook to

```yaml
selfservice:
  methods:
    password:
      config:
        migrate_hook:
          ...
```

jonas-jonas committed:
feat: add tests for two step login (#3959)

aeneasr committed:
chore: regenerate SDK and format code

jonas-jonas committed:
fix: timestamp precision on mysql

aeneasr committed:
chore: rename one_step to unified

jonas-jonas committed:
chore: regenerate config types

jonas-jonas committed:
Update .github/workflows/ci.yaml

Co-authored-by: hackerman <3372410+aeneasr@users.noreply.github.com>

aeneasr committed:
chore: use correct import

jonas-jonas committed:
chore: re-add WithIdentifier (#3992)

jonas-jonas committed:
chore: remove max length

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

hperl committed:
fix: pw migration param (#3998)

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

aeneasr committed:
fix: add continue with only for json browser requests (#4002)

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

hperl committed:
fix: password migration hook config (#4001)

This fixes the config loading for the password migration hook.

ory-bot committed:
autogen(openapi): regenerate swagger spec and internal client

[skip ci]

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

David-Wobrock committed:
fix: typo in login link CLI error messages (#3995)

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

Pedr0Rocha committed:
fix: add fallback to providerLabel (#3999)

This adds a fallback to the provider label when trying to register a duplicate identifier with an oidc.

Current error message:

`Signing in will link your account to "test@test.com" at provider "". If you do not wish to link that account, please start a new login flow.`

The label represents an optional label for the UI, but in my case it's always empty. I suggest we fallback to the provider when the label is not present. In case the label is present, the behaviour won't change.

Fallback to provider:

`Signing in will link your account to "test@test.com" at provider "google". If you do not wish to link that account, please start a new login flow.`

IchordeDionysos committed:
feat: add support for Salesforce as identity provider (#4003)

jonas-jonas committed:
fix: remove flows from log messages (#3913)

Co-authored-by: aeneasr <3372410+aeneasr@users.noreply.github.com>

ory-bot committed:
autogen(openapi): regenerate swagger spec and internal client

[skip ci]

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

IchordeDionysos committed:
deps: update Code QL action to v2 (#4008)

IchordeDionysos committed:
test: update incorrect usage of Auth0 in Salesforce tests (#4007)

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

vinckr committed:
chore: update newsletter link (#4011)

zepatrik committed:
chore: bump dependencies (#4017)

ory-bot committed:
autogen(openapi): regenerate swagger spec and internal client

[skip ci]

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

zepatrik committed:
chore: align internal SDK with published SDK (#4019)

aeneasr committed:
fix: trigger oidc web hook on sign in after registration (#4027)

jonas-jonas committed:
docs: add google to supported providers in ID Token doc strings (#4026)

ory-bot committed:
autogen(openapi): regenerate swagger spec and internal client

[skip ci]

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

zepatrik committed:
fix: refactor internal API to prevent panics (#4028)

ory-bot committed:
autogen(openapi): regenerate swagger spec and internal client

[skip ci]

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

aeneasr committed:
chore: upgrade dockertest to address cve (#4038)

aeneasr committed:
fix: downgrade go-webauthn (#4035)

aeneasr committed:
test: improve stability of refresh test (#4037)

aeneasr committed:
chore: update repository templates to https://github.com/ory/meta/commit/95e82c6db80b6928cfb4a27f165e1ed70d52dee4

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

aeneasr committed:
chore: update repository templates to https://github.com/ory/meta/commit/297c8a5a3563dacb5fb28b60436623ee5608726e

hperl committed:
fix: filter web hook headers (#4048)

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

hperl committed:
fix: concurrent map update for webhook header (#4055)

ory-bot committed:
autogen(openapi): regenerate swagger spec and internal client

[skip ci]

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

jonas-jonas committed:
chore: don't return allowed return URLs (#4044)

hperl committed:
fix: emit SelfServiceMethodUsed in SettingsSucceeded event (#4056)

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

aeneasr committed:
chore: update repository templates to https://github.com/ory/meta/commit/939b80fbfd795b1e137654c60deac78b752e7196

aeneasr committed:
chore: update repository templates to https://github.com/ory/meta/commit/3cf0f005564ff7e18f13640dd3dab677f610049f

aeneasr committed:
chore: update repository templates to https://github.com/ory/meta/commit/4132def510aefccb5bdb55519823bff9b7be5bc7

aeneasr committed:
chore: update repository templates to https://github.com/ory/meta/commit/fe4ffe081358634ef4d59fdc2ffe2741c63981c8

aeneasr committed:
Merge commit from fork

* fix(security): code credential does not respect `highest_available` setting

This patch fixes a security vulnerability which prevents the `code` method to properly report it's credentials count to the `highest_available` mechanism.

For more details on this issue please refer to the [security advisory](https://github.com/ory/kratos/security/advisories/GHSA-wc43-73w7-x2f5).

* fix: normalize code credentials and deprecate via parameter

Before this, code credentials for passwordless and mfa login were incorrectly stored and normalized. This could cause issues where the system would not detect the user's phone number, and where SMS/email MFA would not properly work with the `highest_available` setting.

Breaking changes: Please note that the `via` parameter is deprecated when performing SMS 2FA. It will be removed in a future version. If the parameter is not included in the request, the user will see all their phone/email addresses from which to perform the flow.

Before upgrading, ensure that your identity schema has the appropriate code configuration when using the code method for passwordless or 2fa login.

If you are using the code method for 2FA login already, or you are using it for 1FA login but have not yet configured the code identifier, set `selfservice.methods.code.config.missing_credential_fallback_enabled` to `true` to prevent users from being locked out.

aeneasr committed:
fix: normalize code credentials and deprecate via parameter

Before this, code credentials for passwordless and mfa login were incorrectly stored and normalized. This could cause issues where the system would not detect the user's phone number, and where SMS/email MFA would not properly work with the `highest_available` setting.

Breaking changes: Please note that the `via` parameter is deprecated when performing SMS 2FA. It will be removed in a future version. If the parameter is not included in the request, the user will see all their phone/email addresses from which to perform the flow.

Before upgrading, ensure that your identity schema has the appropriate code configuration when using the code method for passwordless or 2fa login.

If you are using the code method for 2FA login already, or you are using it for 1FA login but have not yet configured the code identifier, set `selfservice.methods.code.config.missing_credential_fallback_enabled` to `true` to prevent users from being locked out.

aeneasr committed:
fix(security): code credential does not respect `highest_available` setting

This patch fixes a security vulnerability which prevents the `code` method to properly report it's credentials count to the `highest_available` mechanism.

For more details on this issue please refer to the [security advisory](https://github.com/ory/kratos/security/advisories/GHSA-wc43-73w7-x2f5).

David-Wobrock committed:
test: resolve CI failures (#4067)

aeneasr committed:
fix: return credentials in FindByCredentialsIdentifier (#4068)

Instead of re-fetching the credentials later (expensive), we load them only once.

jonas-jonas committed:
chore: add missing text message fields (#4066)

jonas-jonas committed:
fix: improve OIDC account linking UI (#4036)

aeneasr committed:
fix: whoami latency (#4070)

zepatrik committed:
fix: validate page tokens for better error codes (#4021)

jonas-jonas committed:
fix: do not populate `id_first` first step for account linking flows (#4074)

jonas-jonas committed:
fix: include duplicate credentials in account linking message (#4079)

alnr committed:
chore: update CI, dependencies, and some inaccurate assertions (#4085)

alnr committed:
feat: client-side PKCE take 3 (#4078)

* feat: client-side PKCE

This change introduces a new configuration for OIDC providers: pkce with values auto (default), never, force.

When auto is specified or the field is omitted, Kratos will perform autodiscovery and perform PKCE when the server advertises support for it. This requires the issuer_url to be set for the provider.

never completely disables PKCE support. This is only theoretically useful: when a provider advertises PKCE support but doesn't actually implement it.

force always sends a PKCE challenge in the initial redirect URL, regardless of what the provider advertises. This setting is useful when the provider offers PKCE but doesn't advertise it in his ./well-known/openid-configuration.

Important: When setting pkce: force, you must whitelist a different return URL for your OAuth2 client in the provider's configuration. Instead of <base-url>/self-service/methods/oidc/callback/<provider>, you must use <base-url>/self-service/methods/oidc/callback (note missing last path segment). This is to enable the use of the same OAuth client ID+secret when configuring several Kratos OIDC providers, without having to whitelist individual redirect_uris for each Kratos provider config.

* chore: regenerate SDK, bump DB versions, cleanup tool install

* chore: get final organization ID from provider config during registration and login

* chore: fixup OIDC function signatures and improve tests

zepatrik committed:
chore: improve tracing for selfservice strategies (#4061)

alnr committed:
fix: add PKCE config key to config schema (#4098)

hperl committed:
chore: sdk+ci (#4088)

dependabot[bot] committed:
chore(deps): bump github.com/opencontainers/runc from 1.1.13 to 1.1.14 (#4075)

Bumps [github.com/opencontainers/runc](https://github.com/opencontainers/runc) from 1.1.13 to 1.1.14.
- [Release notes](https://github.com/opencontainers/runc/releases)
- [Changelog](https://github.com/opencontainers/runc/blob/main/CHANGELOG.md)
- [Commits](https://github.com/opencontainers/runc/compare/v1.1.13...v1.1.14)

---
updated-dependencies:
- dependency-name: github.com/opencontainers/runc
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
chore(deps): bump express from 4.18.2 to 4.20.0 in /test/e2e/proxy (#4095)

Bumps [express](https://github.com/expressjs/express) from 4.18.2 to 4.20.0.
- [Release notes](https://github.com/expressjs/express/releases)
- [Changelog](https://github.com/expressjs/express/blob/master/History.md)
- [Commits](https://github.com/expressjs/express/compare/4.18.2...4.20.0)

---
updated-dependencies:
- dependency-name: express
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
chore(deps): bump body-parser and express in /test/e2e/proxy (#4093)

Bumps [body-parser](https://github.com/expressjs/body-parser) to 1.20.3 and updates ancestor dependency [express](https://github.com/expressjs/express). These dependencies need to be updated together.


Updates `body-parser` from 1.20.1 to 1.20.3
- [Release notes](https://github.com/expressjs/body-parser/releases)
- [Changelog](https://github.com/expressjs/body-parser/blob/master/HISTORY.md)
- [Commits](https://github.com/expressjs/body-parser/compare/1.20.1...1.20.3)

Updates `express` from 4.18.2 to 4.20.0
- [Release notes](https://github.com/expressjs/express/releases)
- [Changelog](https://github.com/expressjs/express/blob/master/History.md)
- [Commits](https://github.com/expressjs/express/compare/4.18.2...4.20.0)

---
updated-dependencies:
- dependency-name: body-parser
  dependency-type: indirect
- dependency-name: express
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
chore(deps): bump serve-static and express in /test/e2e/proxy (#4091)

Bumps [serve-static](https://github.com/expressjs/serve-static) to 1.16.0 and updates ancestor dependency [express](https://github.com/expressjs/express). These dependencies need to be updated together.


Updates `serve-static` from 1.15.0 to 1.16.0
- [Release notes](https://github.com/expressjs/serve-static/releases)
- [Changelog](https://github.com/expressjs/serve-static/blob/master/HISTORY.md)
- [Commits](https://github.com/expressjs/serve-static/compare/v1.15.0...1.16.0)

Updates `express` from 4.18.2 to 4.20.0
- [Release notes](https://github.com/expressjs/express/releases)
- [Changelog](https://github.com/expressjs/express/blob/master/History.md)
- [Commits](https://github.com/expressjs/express/compare/4.18.2...4.20.0)

---
updated-dependencies:
- dependency-name: serve-static
  dependency-type: indirect
- dependency-name: express
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

splaunov committed:
fix: transient_payload is lost when verification flow started as part of registration (#3983)

aeneasr committed:
fix: incorrect append of code credential identifier (#4102)

Closes #4076

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

alnr committed:
chore: refactor API in package cipher for easier dependency injection (#4103)

ory-bot committed:
autogen(openapi): regenerate swagger spec and internal client

[skip ci]

aeneasr committed:
feat: emit events in identity persister (#4107)

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

hperl committed:
feat: allow partially failing batch inserts (#4083)

When batch-inserting multiple identities, conflicts or validation errors of a subset of identities in the batch still allow the rest of the identities to be inserted. The returned JSON contains the error details that lead to the failure.

---------

Co-authored-by: Patrik <zepatrik@users.noreply.github.com>

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

hperl committed:
fix: batch identity created event (#4111)

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

jonas-jonas committed:
fix: return error if invalid UUID is supplied to ids filter (#4116)

alnr committed:
feat: enable new-style OIDC state generation (#4121)

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

aeneasr committed:
chore: upgrade goreleaser to v2 (#4123)

aeneasr committed:
autogen: pin v1.3.0-pre.0 release commit

aeneasr committed:
test: additional code credential test case (#4122)

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

aeneasr committed:
feat: change `method=profile:back` to `screen=previous` (#4119)

BREAKING CHANGE: When using two-step registration, it was previously possible to send `method=profile:back` to get to the previous screen. This feature was not documented in the SDK API yet. Going forward, please instead use `screen=previous`.

ory-bot committed:
autogen(openapi): regenerate swagger spec and internal client

[skip ci]

ory-bot committed:
autogen(docs): regenerate and update changelog

[skip ci]

alnr committed:
fix: passthrough correct organization ID to CompletedLoginForWithProvider (#4124)

aeneasr committed:
autogen: pin v1.3.0 release commit

aeneasr committed:
feat: support android webauthn origins

This patch adds the ability to verify Android APK origins used during WebAuthn/Passkey exchange.

Upgrades go-webauthn and includes fixes for Go 1.23 and workarounds for Swagger.

aeneasr committed:
autogen: pin v1.3.1 release commit

maoanran committed:
Squashed commit of the changes before v1.3.1:

commit 8b706f5fc3242bb01418608f1b682c7bb2fd4f38
Author: splaunov <splaunov@gmail.com>
Date:   Wed Dec 4 05:19:46 2024 +0300

    feat: support registration flow via SAML (PS-106)

commit ba9635dccfb977c1ce71dcfc7b933457e6f7df7f
Author: splaunov <splaunov@gmail.com>
Date:   Mon Nov 4 10:46:19 2024 +0300

    feat: add duplicate credentials error handling to saml api flow (PS-567)

commit 25fdf42e3ce57c558edcc9e9233bb21a4b39e464
Author: splaunov <splaunov@gmail.com>
Date:   Fri Jun 28 22:17:09 2024 +0300

    feat: fix sso(saml) e2e test (PS-366)

commit e9b2e38cd5848062302a1384142179dc264b02b4
Author: splaunov <splaunov@gmail.com>
Date:   Thu Jun 27 18:50:55 2024 +0300

    feat: do not use cookies to save state in SAML flow (PS-15)

commit 9ef0a8abd80833e7db4165b03e5f1a00fea1b47c
Author: ThibaultHerard <thibaultherard10@gmail.com>
Date:   Tue Aug 30 15:12:58 2022 +0000

    all commits before 2024 rebase

    feat(saml): api flow support (PS-35)

    all commits before Jan 2024

    fix(saml): applied recent changes from ovh/saml branch (PS-35)

    fix(saml): fixed compilation errors after rebase (PS-35)

    feat: extend admin api to read SAML providers configs (CORE-966)

    (cherry picked from commit a55dee27fba17fbb208d9c07455f66f6bad365ad)

    ignore: add test to check token is valid in webview flow (CORE-2202)

    (cherry picked from commit 2cee177bc4a0b656107b671aff300e5e9e3ba475)

    feat(saml): add “provider id” parameter to kratos session (CORE-2073)

    (cherry picked from commit ffa32d7b27fc032c49c80ce5b2d933837aa9f529)

    feat(oidc): add “provider id” parameter to kratos session (CORE-2073)

    (cherry picked from commit ddea2693636fe75fe1c608fa40538b3fdd7f6803)

    feat: link credentials when second login is OIDC (CORE-2041)

    (cherry picked from commit 2a0b706cb607f994e3f3c21e190563789f75d9be)

    feat: extend SAML cookie lifetime and set SameSite to None (CORE-2024)

    (cherry picked from commit ef9918f8cabaf1c40a621e01807caa0319a71b5e)

    feat: extend SAML cookie lifetime and set SameSite to None (CORE-2024)

    (cherry picked from commit 7c6352a51aebd55cdcf327520b7f3bf5211f74af)

    fix: panic when linked flow does not exist (CORE-2017)

    (cherry picked from commit 98597ef620fa8e127a2158fa4bab54af354a39bc)

    feat(saml): return flow id when doing sso webview first login and email already exists (CORE-1986)

    (cherry picked from commit 10690c8523e9acd9bacd74b048325cf88c2e470a)

    feat(saml): return error in webview redirect url when provider is not found (CORE-1982)

    (cherry picked from commit 58be1dfe1e6e9ee26b1b73e719be3db03b0e86b7)

    fix(saml): missed CSRF token when login with Google SAML and email is already used (CORE-1975)

    (cherry picked from commit d808af6d6394164647aab43c11171b9251e690a8)

    feat: link credentials when login - add LoginAndLinkCredentials method to login and registrations flows

    (cherry picked from commit a6b8ad18b61afec46b667073c250c091dd38ff91)

    feat: link credentials when login

    (cherry picked from commit 9ea27079cfe777239e3095152a422d11e7abd464)

    fix(saml): saml settings link redirect path

    (cherry picked from commit 514b47bd6114c6bcf6c45d668d52c1473621b33b)

    feat(saml): saml link/unlink

    (cherry picked from commit 2cf82e822cd267fdf190ae5fd9391adfcdaa4149)

    fix(saml): forward to registration flow when user with email is found

    (cherry picked from commit 8f145c80c227c7fd3a194f20566c78c8130c6dc9)

    fix(saml): forward to registration flow when user with email is found

    (cherry picked from commit 58165d50ea3307df506c88919cf5a21f74ec7b38)

    fix(saml): An error occurred while retrieving the middeware, it is null

    (cherry picked from commit c00246af21943b1a9c849e966808180ab30b4cbf)

    fix(saml): saml creates identity with wrong default schema id

    (cherry picked from commit 995d429d0bc1309b35ed1075b05b0d36a5cb9367)

    fix(saml): add debug info to errors

    (cherry picked from commit ef7fab243ab73906621cbfedda935e99abf378d9)

    fix(saml): add debug info to errors

    (cherry picked from commit e836c7495b7fc6908a67b4297f3da221e84b442a)

    feat(saml): set login flow error code and message if provider not found

    (cherry picked from commit 2ac84f4c5942ff9267b5082af60031821de1753d)

    feat(saml): support WebView flow for saml strategy

    (cherry picked from commit dceaea1625337a05e719cba86620608fe89937ed)

    feat(saml): relaystate continuity fix + unit tests
    Signed-off-by: sebferrer <sebferrer@users.noreply.github.com>

    Co-authored-by: ThibaultHerard <thibaultherard10@gmail.com>
    (cherry picked from commit aeb172c630d280e24e73dc0d83aeaac0816de169)

    feat(saml): improved error handling
    Signed-off-by: ThibaultHerard <thibaultherard10@gmail.com>

    Co-authored-by: sebferrer <sebferrer@users.noreply.github.com>
    (cherry picked from commit 37f7bbe433dfee5ad65aed13b58e32595f9e4fc6)

    feat(saml): use ory/x fetcher
    Signed-off-by: ThibaultHerard <thibaultherard10@gmail.com>

    Co-authored-by: sebferrer <sebferrer@users.noreply.github.com>
    (cherry picked from commit b26f11d3961a0bf342d1e3a2431db129278055dc)

    feat(saml): saml 2.0 implementation
    Signed-off-by: ThibaultHerard <thibaultherard10@gmail.com>

    Co-authored-by: sebferrer <sebferrer@users.noreply.github.com>
    Co-authored-by: psauvage <psauvage0@users.noreply.github.com>
    Co-authored-by: alexGNX <alexGNX@users.noreply.github.com>
    Co-authored-by: Stoakes <Stoakes@users.noreply.github.com>
    (cherry picked from commit 3f5480fef6bf7aa3c24ac8a1fd2309c3e1c115ec)
